### ISSUE TITLE
Unnaturally high pending_favs

### ISSUE THREAD

**Issue by mlgiraud** [aflnet/aflnet#20]
In order to have a baseline to compare against for my thesis i implemented the OPC UA protocol for aflnet.
It seems to work so far, but the pending_favs variable takes on some unnaturally high values.
See the following images for example:
![image](https://user-images.githubusercontent.com/1338425/91965822-54229180-ed11-11ea-8af1-d6cefc3ef876.png)

Have you come across this issue as well? I'm not sure if this is from my implementation of the protocol, or maybe a bug in the stateful code.
The relevant parts should be on my fork now: https://github.com/mlgiraud/aflnet-1/tree/feature/opcua

**Comment by thuanpv** [comment]
Thanks, @mlgiraud. I think it might be a bug in AFLNet.

**Comment by thuanpv** [comment]
Hi @mlgiraud. It would be very helpful if you could share with me a Dockerfile with everything ready to reproduce the issue.

**Comment by mlgiraud** [comment]
@thuanpv I can try to create a dockerfile as soon as i have some more time.
Also, as soon as i get to that i will most likely get into touch with you guys again regarding the state identification, as i am starting to see some results there and would like to share and discuss with you. I will contact you via email regarding that in the next weeks.

On a sidenote:
Did you observe a low percentage of testcases with new edges in your experiments as well? After one hour of fuzzing only around 23% of the testcases produce new edge coverage. The rest only exercises previously discovered edges a different number of times.

PS: I think this might be an integer underflow bug. I encountered a similar problem with my implementation of the state identification.

**Comment by mlgiraud** [comment]
I created a docker image that runs aflnet with the target application, but i cannot seem to reproduce the issue anymore.
Although now there is a different issue, where the test crashes after a few hours because the testfile could not be written correctly.
I checked if it is a space problem on my side, but that is very unlikely, since there was more than 7GB on the ramdisk remaining.

You can find the docker image here: https://data.mgiraud.de/github-shared/aflnet/aflnet-open62541.tar
After loading the image you can run it with `docker run -d -t --user ${UID} -v your_local_out_path:/out --name AFLNET aflnet:latest`.
Remember to adapt the output path to an appropriate location.
You can also override the cmd to launch a shell and attach directly. Then you can start the fuzzer manually.
The default command ist `/resources/afl-fuzz -t 1000 -d -i /resources/corpus -o out -N tcp://127.0.0.1/4840 -P OPCUA -D 10000 -q 3 -s 3 -E -K -R /resources/target`.

**Comment by thuanpv** [comment]
Thanks a lot, @mlgiraud. I am running it to reproduce the "short write to (null)" error. I did get this error in the past even with the normal AFL. I remember it was due to limited free disk space because I did not use ramdisk at that time.

**Comment by mlgiraud** [comment]
So it might be that the mutations produce a really large input and that crashes the fuzzer.
Shouldn't the size of the generated seed be limited somehow?

**Comment by thuanpv** [comment]
Yes, I think this is a good idea. In this recently published paper (https://www.mdpi.com/1424-8220/20/18/5194), the authors report a similar observation -- AFLNet generates some really large inputs. 

Actually AFLNet doesn't need that file at all because the messages are stored in memory so I was considering to ignore the write_to_testcase function. However, there were some dependencies preventing me from doing so. I will revisit the code and try to come up with a good solution. 

BTW, AFLNet has not crashes yet on my computer after 20+ hours. All metrics look ok (e.g., 1489 paths and a very high stability at 94.52%).

**Comment by mlgiraud** [comment]
How many executions per second are you running at? After 15 hours i was at 1M total executions for reference.

**Comment by thuanpv** [comment]
Following is the statistics shown in fuzzer_stats

start_time        : 1603252850
last_update       : 1603337146
fuzzer_pid        : 11
cycles_done       : 259
execs_done        : 1447740
execs_per_sec     : 10.33
paths_total       : 1519
paths_favored     : 78
paths_found       : 1504
paths_imported    : 0
max_depth         : 26
cur_path          : 1073
pending_favs      : 2
pending_total     : 857
variable_paths    : 126
stability         : 94.52%
bitmap_cvg        : 10.19%
unique_crashes    : 0
unique_hangs      : 0
last_path         : 1603336940
last_crash        : 0
last_hang         : 0
execs_since_crash : 1447740
exec_timeout      : 1000
afl_banner        : target
afl_version       : 2.56b
target_mode       : default
command_line      : /resources/afl-fuzz -t 1000 -d -i /resources/corpus -o out -N tcp://127.0.0.1/4840 -P OPCUA -D 10000 -q 3 -s 3 -E -K -R /resources/target
slowest_exec_ms   : 668

```
If you like, you can send me the results of your run. I would be interested looking at them.
```
Sure, would the whole output folder (/out) be enough? It is ~300MB.

**Comment by mlgiraud** [comment]
Ok so your run managed to get about 400k more iterations. Did it crash, or did you cancel it after this time?
I'm going to run a new evaluation over the weekend in the hopes it will be able to run for the 48 hours.

**Comment by thuanpv** [comment]
No, it did not crash; I stopped it. From next week, I will spend more time to fix bugs/improve the tool.

**Comment by shtry** [comment]
same error.

#31

```
[-] PROGRAM ABORT : Short write to (null)
         Location : write_to_testcase(), afl-fuzz.c:3347
```

**Comment by acidghost** [comment]
@mlgiraud Have you found a stable solution to the `Short write to (null)` error?
